[MediaStream] Add support for ImageCapture.takePhoto

eric-carlson opened the pull request:
#### 6b87ac816d8acb2ee9016d88e7a5d639b19e5818
<pre>
[MediaStream] Add support for ImageCapture.takePhoto
<a href="https://bugs.webkit.org/show_bug.cgi?id=262467">https://bugs.webkit.org/show_bug.cgi?id=262467</a>
<a href="https://rdar.apple.com/116322637">rdar://116322637</a>

Reviewed by Youenn Fablet.

Implement ImageCapture.takePhoto in AVVideoCaptureSource and MockRealtimeVideoSource.
We reconfigure the capture source when asked for a photo larger than the current preset,
so it always returns an image at least as large as the requested size, but it doesn&apos;t
yet resize the image if the requested size doesn&apos;t exactly match the preset.

Taking a photo is an asynchronous operation that can take an arbitrary amount of time, so
we don&apos;t do it on the main thread. Because we may need to reconfigure the capture device
to take a photo with the requested settings, we need to prevent the capture device from
being reconfigured by `applyConstraints` while a photo is being generated. To do this we
serialize calls to `takePhoto` and `applyConstraints` so they don&apos;t interfere with one
another.

* LayoutTests/fast/mediastream/image-capture-take-photo-expected.txt: Added.
* LayoutTests/fast/mediastream/image-capture-take-photo.html: Added.
* Source/WebCore/Modules/mediastream/ImageCapture.cpp:
(WebCore::ImageCapture::takePhoto):
* Source/WebCore/Modules/mediastream/ImageCapture.h:
* Source/WebCore/Modules/mediastream/ImageCapture.idl:
* Source/WebCore/Modules/mediastream/MediaStreamTrack.cpp:
(WebCore::MediaStreamTrack::stopTrack):
(WebCore::MediaStreamTrack::queueAndProcessSerialAction):
(WebCore::MediaStreamTrack::processNextSerialAction):
(WebCore::MediaStreamTrack::takePhoto):
(WebCore::MediaStreamTrack::applyConstraints):
* Source/WebCore/Modules/mediastream/MediaStreamTrack.h:
* Source/WebCore/PAL/pal/cocoa/AVFoundationSoftLink.h:
* Source/WebCore/PAL/pal/cocoa/AVFoundationSoftLink.mm:
* Source/WebCore/platform/mediastream/MediaStreamTrackPrivate.cpp:
(WebCore::MediaStreamTrackPrivate::takePhoto):
* Source/WebCore/platform/mediastream/MediaStreamTrackPrivate.h:
* Source/WebCore/platform/mediastream/PhotoSettings.h:
(WebCore::operator==):
* Source/WebCore/platform/mediastream/RealtimeMediaSource.cpp:
(WebCore::RealtimeMediaSource::takePhoto):
* Source/WebCore/platform/mediastream/RealtimeMediaSource.h:
* Source/WebCore/platform/mediastream/RealtimeVideoCaptureSource.cpp:
(WebCore::RealtimeVideoCaptureSource::bestSupportedSizeFrameRateAndZoomConsideringObservers):
(WebCore::RealtimeVideoCaptureSource::setSizeFrameRateAndZoom):
(WebCore::RealtimeVideoCaptureSource::takePhotoInternal):
(WebCore::RealtimeVideoCaptureSource::takePhoto):
* Source/WebCore/platform/mediastream/RealtimeVideoCaptureSource.h:
* Source/WebCore/platform/mediastream/mac/AVVideoCaptureSource.h:
* Source/WebCore/platform/mediastream/mac/AVVideoCaptureSource.mm:
(WebCore::toFillLightMode):
(WebCore::toAVCaptureFlashMode):
(WebCore::photoQueue):
(WebCore::AVVideoCaptureSource::~AVVideoCaptureSource):
(WebCore::AVVideoCaptureSource::stopProducingData):
(WebCore::AVVideoCaptureSource::stopSession):
(WebCore::AVVideoCaptureSource::photoOutput):
(WebCore::AVVideoCaptureSource::resolvePendingPhotoRequest):
(WebCore::AVVideoCaptureSource::rejectPendingPhotoRequest):
(WebCore::AVVideoCaptureSource::maxPhotoSizeForCurrentPreset const):
(WebCore::AVVideoCaptureSource::photoConfiguration):
(WebCore::AVVideoCaptureSource::takePhotoInternal):
(WebCore::AVVideoCaptureSource::getPhotoSettings):
(WebCore::AVVideoCaptureSource::captureOutputDidFinishProcessingPhoto):
(-[WebCoreAVVideoCaptureSourceObserver captureOutput:didFinishProcessingPhoto:error:]):
* Source/WebCore/platform/mediastream/mac/MockRealtimeVideoSourceMac.mm:
(WebCore::MockRealtimeVideoSourceMac::updateSampleBuffer):
* Source/WebCore/platform/mock/MockRealtimeVideoSource.cpp:
(WebCore::takePhotoRunLoop):
(WebCore::MockRealtimeVideoSource::DrawingState::fontDescription):
(WebCore::MockRealtimeVideoSource::DrawingState::timeFont):
(WebCore::MockRealtimeVideoSource::DrawingState::bipBopFont):
(WebCore::MockRealtimeVideoSource::DrawingState::statsFont):
(WebCore::MockRealtimeVideoSource::takePhotoInternal):
(WebCore::MockRealtimeVideoSource::invalidateDrawingState):
(WebCore::MockRealtimeVideoSource::drawingState):
(WebCore::MockRealtimeVideoSource::settingsDidChange):
(WebCore::MockRealtimeVideoSource::drawText):
(WebCore::MockRealtimeVideoSource::generatePhoto):
(WebCore::MockRealtimeVideoSource::generateFrameInternal):
(WebCore::MockRealtimeVideoSource::generateFrame):
(WebCore::MockRealtimeVideoSource::imageBuffer):
(WebCore::MockRealtimeVideoSource::imageBufferInternal):
(WebCore::MockRealtimeVideoSource::imageBuffer const): Deleted.
* Source/WebCore/platform/mock/MockRealtimeVideoSource.h:
(WebCore::MockRealtimeVideoSource::DrawingState::DrawingState):
(WebCore::MockRealtimeVideoSource::DrawingState::baseFontSize const):
(WebCore::MockRealtimeVideoSource::DrawingState::statsFontSize const):
* Source/WebKit/GPUProcess/webrtc/RemoteSampleBufferDisplayLayer.h:
* Source/WebKit/UIProcess/Cocoa/UserMediaCaptureManagerProxy.cpp:
(WebKit::UserMediaCaptureManagerProxy::SourceProxy::takePhoto):
(WebKit::UserMediaCaptureManagerProxy::takePhoto):
* Source/WebKit/UIProcess/Cocoa/UserMediaCaptureManagerProxy.h:
* Source/WebKit/UIProcess/Cocoa/UserMediaCaptureManagerProxy.messages.in:
* Source/WebKit/WebProcess/cocoa/RemoteRealtimeMediaSource.cpp:
(WebKit::RemoteRealtimeMediaSource::takePhoto):
* Source/WebKit/WebProcess/cocoa/RemoteRealtimeMediaSource.h:
* Source/WebKit/WebProcess/cocoa/RemoteRealtimeMediaSourceProxy.cpp:
(WebKit::RemoteRealtimeMediaSourceProxy::takePhoto):
* Source/WebKit/WebProcess/cocoa/RemoteRealtimeMediaSourceProxy.h:
ef8342e

Canonical link: <a href="https://commits.webkit.org/270621@main">https://commits.webkit.org/270621@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/dc92712f660302ccc44bc11381fadd92a3d1cb3e

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/25944 "4 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/4555 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/27225 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/28044 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/23759 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/26262 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/6317 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/1983 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/23838 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/26196 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/3437 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/22352 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/28624 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/3057 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/23305 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/29374 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/23671 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/23682 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/27250 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/3089 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/1298 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/4479 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/23053 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6234 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/3546 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/3408 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

youennf commented on Source/WebCore/Modules/mediastream/MediaStreamTrack.cpp:
```diff
@@ -262,6 +264,10 @@ void MediaStreamTrack::stopTrack(StopMode mode)
 
     configureTrackRendering();
+
+    auto pendingActions = WTFMove(m_pendingActionQueue);
```
std:exchange would be better.

youennf commented on Source/WebCore/Modules/mediastream/MediaStreamTrack.cpp:
```diff
@@ -262,6 +264,10 @@ void MediaStreamTrack::stopTrack(StopMode mode)
+
+    auto pendingActions = WTFMove(m_pendingActionQueue);
+    if (!pendingActions.isEmpty())
+        pendingActions.takeFirst()(SerialActionResolution::Cancel);
```
Why only the first action?

jyavenard replied:
> Why only the first action?

it's dequeuing in the loop.

youennf commented on Source/WebCore/Modules/mediastream/MediaStreamTrack.cpp:
```diff
@@ -314,6 +320,75 @@ MediaStreamTrack::TrackCapabilities MediaStreamTrack::getCapabilities() const
+    TakePhotoPromise::Producer producer;
+    Ref<TakePhotoPromise> promise = producer;
+
+    queueAndProcessSerialAction([settings = WTFMove(settings), protectedThis = Ref { *this }, producer = WTFMove(producer)](SerialActionResolution resolution) mutable -> Ref<GenericPromise> {
```
It might be good to explain in the commit message why we are introducing a queue.

youennf commented on Source/WebCore/Modules/mediastream/MediaStreamTrack.h:
```diff
@@ -202,6 +203,14 @@ class MediaStreamTrack
+
+    Deque<SerialAction> m_pendingActionQueue;
+    std::optional<SerialAction> m_currentSerialAction;
+    bool m_executingSerialAction { false };
```
Not used

youennf commented on Source/WebCore/Modules/mediastream/MediaStreamTrack.cpp:
```diff
@@ -314,6 +320,75 @@ MediaStreamTrack::TrackCapabilities MediaStreamTrack::getCapabilities() const
+    if (m_currentSerialAction || m_pendingActionQueue.isEmpty())
+        return;
+
+    m_currentSerialAction = { m_pendingActionQueue.takeFirst() };
```
It is not clear to me whether we need a m_pendingActionQueue.
Maybe what we need is using chainTo, let's check with Jean-Yves

jyavenard replied:
The queue is used for two types of task: applyConstraints which is using completionHandler and takePhoto that is returning a promise.
You can't take a photo unless applyConstraints is done.

if applyConstraints returned a NativePromise, we could indeed just chain the promises together, so that takePhoto would automatically occurs when applyConstraints is finished. 

We could probably rather than using a dedicated queue chain the action. End result would be the same

youennf replied:
I am not sure we need to wait for applyConstraints to call takePhoto.
First, normally, applyConstraints succeeds synchronously in GPUProcess, we just have to ensure that the order to applyConstraints is not following a different path than the order to takePhoto.
Second, applyConstraints is tied to a track, but you can call several applyConstraints on different tracks pointing to the same GPU source (either via track.clone() or calling getUserMedia twice).
If we have to create a queue, it would best be handled in GPUProcess by the capture source itself.

youennf commented on Source/WebCore/Modules/mediastream/ImageCapture.cpp:
```diff
@@ -53,6 +54,16 @@ ImageCapture::ImageCapture(Document& document, Ref<MediaStreamTrack> track)
+        if (!result)
+            promise.reject(WTFMove(result.error()));
+        else
+            promise.resolve(Blob::create(protectedThis->scriptExecutionContext(), WTFMove(get<0>(result.value())), WTFMove(get<1>(result.value()))));
```
We want to resolve the promise in an event task queue, which is not guaranteed here.

Maybe we actually want to specialise SerialFunctionDispatcher for ScriptExecutionContext (or maybe Document) so that we ensure that the callback is called within an event task.
It would also write better since we would not use `RunLoop::main()` but `scriptExecutionContext ()`

youennf commented on Source/WebCore/Modules/mediastream/MediaStreamTrack.cpp:
```diff
@@ -314,6 +320,75 @@ MediaStreamTrack::TrackCapabilities MediaStreamTrack::getCapabilities() const
+                else if (RefPtr context = protectedThis->scriptExecutionContext(); !context || context->activeDOMObjectsAreStopped() || protectedThis->m_ended)
+                    producer.reject(Exception { OperationError, "Track has ended"_s });
+                else
+                    producer.resolve(WTFMove(result.value()));
```
I would tend to create the blob here instead of the call site.

jyavenard replied:
I suggested creating the blob at the point where the DOMPromise is resolved, as it clearly separate the JS side of things from the implementation.

Also allows to avoid to unnecessarily link all JS headers that you don't have to use.

To me, resolving the DOM promise at the same level the DOMPromise was first received makes for clearer reading/review.

jyavenard commented on Source/WebCore/Modules/mediastream/MediaStreamTrack.cpp:
```diff
@@ -35,6 +35,7 @@
 #include "Event.h"
 #include "EventNames.h"
 #include "FrameLoader.h"
+#include "JSBlob.h"
```
unused

jyavenard commented on Source/WebCore/Modules/mediastream/MediaStreamTrack.cpp:
```diff
@@ -61,6 +62,7 @@
 #include <wtf/IsoMallocInlines.h>
 #include <wtf/NativePromise.h>
 #include <wtf/NeverDestroyed.h>
+#include <wtf/Scope.h>
```
used?

youennf commented on Source/WebCore/Modules/mediastream/ImageCapture.h:
```diff
@@ -28,6 +28,7 @@
 #if ENABLE(MEDIA_STREAM)
 
 #include "ActiveDOMObject.h"
+#include "Blob.h"
```
I am not sure we need ImageCapture to be an ActiveDOMObject, a ContextDestructionObserver might be sufficient, we could look at this in a follow-up

youennf commented on Source/WebCore/Modules/mediastream/ImageCapture.idl:
```diff
@@ -34,8 +34,7 @@
-    // FIXME: https://bugs.webkit.org/show_bug.cgi?id=262467
-    // Promise<Blob> takePhoto(optional PhotoSettings photoSettings = {});
+    [NewObject] Promise<Blob> takePhoto(optional PhotoSettings photoSettings = {});
 
```
NewObject is probably also in getPhotoSettings?

youennf commented on Source/WebCore/Modules/mediastream/MediaStreamTrack.cpp:
```diff
@@ -314,6 +314,37 @@ MediaStreamTrack::TrackCapabilities MediaStreamTrack::getCapabilities() const
+    // If the readyState of track provided in the constructor is not live, return
+    // a promise rejected with a new DOMException whose name is InvalidStateError,
+    // and abort these steps.
+    if (m_ended || m_readyState != State::Live)
```
It is a bit odd to have that handling in MediaStreamTrack::takePhoto but not in MediaStreamTrack::getPhotoCapabilities. If we move them all to MediaStreamTrack, it will be easier to keep `Track has ended` string in sync.

m_ended should be sufficient also, no?
If not, we probably want it in step 339.

youennf commented on Source/WebCore/Modules/mediastream/MediaStreamTrack.cpp:
```diff
@@ -314,6 +314,37 @@ MediaStreamTrack::TrackCapabilities MediaStreamTrack::getCapabilities() const
+        // abort these steps.
+        if (!result)
+            producer.reject(Exception { UnknownError, WTFMove(result.error()) });
+        else if (RefPtr context = protectedThis->scriptExecutionContext(); !context || context->activeDOMObjectsAreStopped() || protectedThis->m_ended)
```
Could use if () { return; } to remove the `else if`

youennf commented on Source/WebCore/Modules/mediastream/MediaStreamTrack.cpp:
```diff
@@ -367,15 +398,20 @@ static MediaConstraints createMediaConstraints(const std::optional<MediaTrackCon
+        return;
+    }
+
+    m_private->applyConstraints(createMediaConstraints(constraints), [protectedThis = Ref { *this }, constraints, promise = WTFMove(promise)](auto&& error) mutable {
```
This is a preexisting issue but we should probably queue a task here like we are doing in ImageCapture::takePhoto.

youennf commented on Source/WebCore/platform/mediastream/RealtimeVideoCaptureSource.cpp:
```diff
@@ -412,20 +412,143 @@ void RealtimeVideoCaptureSource::setSizeFrameRateAndZoom(std::optional<int> widt
+    TakePhotoNativePromise::Producer takePhotoProducer;
+    Ref<TakePhotoNativePromise> takePhotoPromise = takePhotoProducer;
+
+    queueAndProcessSerialAction([photoSettings = WTFMove(photoSettings), protectedThis = Ref { *this }, takePhotoProducer = WTFMove(takePhotoProducer), this]() mutable -> Ref<GenericPromise> {
```
Maybe put `this` closer to `protectedThis`.

youennf commented on Source/WebCore/platform/mediastream/RealtimeVideoCaptureSource.cpp:
```diff
@@ -412,20 +412,143 @@ void RealtimeVideoCaptureSource::setSizeFrameRateAndZoom(std::optional<int> widt
+            // streaming SHOULD cause onmute and onunmute events to fire on the track in question.
+            if (!muted()) {
+                setMuted(true);
+                m_mutedForPhotoCapture = true;
```
One potential issue is that this will make our UI button flaky for a bit, while we are not stopping capture.
I do not have a good suggestion though, I am tempted to leave this signalling aside for now.

youennf commented on Source/WebCore/platform/mediastream/mac/AVVideoCaptureSource.mm:
```diff
@@ -915,6 +1105,22 @@ static inline IntDegrees sensorOrientation(AVCaptureVideoOrientation videoOrient
+{
+    if (!error) {
+        NSData* data = [photo fileDataRepresentation];
+        resolvePendingPhotoRequest({ static_cast<const uint8_t*>(data.bytes), data.length }, "image/jpeg"_s);
```
return here and no `else`.
I would tend to do `if (error)` as well instead of `if(!error)`

youennf commented on Source/WebCore/platform/mock/MockRealtimeVideoSource.h:
```diff
@@ -107,11 +112,39 @@ class MockRealtimeVideoSource : public RealtimeVideoCaptureSource, private Orien
-    float m_statsFontSize { 0 };
+    class DrawingState {
+    public:
+        DrawingState(float baseFontSize)
```
explicit

youennf commented on Source/WebKit/UIProcess/Cocoa/UserMediaCaptureManagerProxy.cpp:
```diff
@@ -476,18 +486,27 @@ void UserMediaCaptureManagerProxy::createMediaSourceForCaptureDeviceWithConstrai
 
-    if (constraints && proxy->source().type() == RealtimeMediaSource::Type::Video) {
-        if (auto result = proxy->applyConstraints(*constraints)) {
+    auto completeSetup = [this](std::unique_ptr<SourceProxy>&& proxy, RealtimeMediaSourceIdentifier id, CreateSourceCallback&& completionHandler) mutable {
```
Let's take a WeakPointer here.

youennf commented on Source/WebKit/UIProcess/Cocoa/UserMediaCaptureManagerProxy.cpp:
```diff
@@ -535,12 +554,14 @@ void UserMediaCaptureManagerProxy::applyConstraints(RealtimeMediaSourceIdentifie
-        m_connectionProxy->connection().send(Messages::UserMediaCaptureManager::ApplyConstraintsFailed(id, result->badConstraint, result->message), 0);
-        return;
-    }
+    proxy->applyConstraints(constraints, [this, id, proxy](auto&& result) {
```
WeakPtr here

youennf commented on Source/WebKit/UIProcess/Cocoa/UserMediaCaptureManagerProxy.cpp:
```diff
@@ -561,6 +582,21 @@ void UserMediaCaptureManagerProxy::clone(RealtimeMediaSourceIdentifier clonedID,
+        return;
+    }
+
+    proxy->takePhoto(WTFMove(settings))->whenSettled(RunLoop::main(), [handler = WTFMove(handler)] (auto&& result) mutable {
```
Can we write it as `whenSettled(RunLoop::main(), WTFMove(handler))`

youennf commented on Source/WebCore/platform/mediastream/RealtimeVideoCaptureSource.cpp:
```diff
@@ -412,20 +412,143 @@ void RealtimeVideoCaptureSource::setSizeFrameRateAndZoom(std::optional<int> widt
+
+void RealtimeVideoCaptureSource::applyConstraints(const MediaConstraints& constraints, ApplyConstraintsHandler&& callback)
+{
+    queueAndProcessSerialAction([protectedThis = Ref { *this }, &constraints, callback = WTFMove(callback), this]() mutable -> Ref<GenericPromise> {
```
I would move this closer to protectedThis.

youennf commented on Source/WebCore/platform/mediastream/RealtimeVideoCaptureSource.h:
```diff
@@ -81,6 +79,9 @@ class WEBCORE_EXPORT RealtimeVideoCaptureSource : public RealtimeMediaSource, pu
 
     static std::span<const IntSize> standardVideoSizes();
 
+    virtual Ref<TakePhotoNativePromise> takePhotoInternal(PhotoSettings&&);
```
Can be private?

youennf commented on Source/WebCore/platform/mediastream/RealtimeVideoCaptureSource.cpp:
```diff
@@ -412,20 +412,143 @@ void RealtimeVideoCaptureSource::setSizeFrameRateAndZoom(std::optional<int> widt
+            if (result)
+                takePhotoProducer.resolve(WTFMove(result.value()));
+            else
+                takePhotoProducer.reject(WTFMove(result.error()));
```
Can we use `settle` here?

youennf commented on Source/WebCore/platform/mediastream/mac/AVVideoCaptureSource.mm:
```diff
@@ -448,6 +514,146 @@ static bool isZoomSupported(const Vector<VideoPreset>& presets)
+    {
+        Locker lock { m_photoLock };
+        if (m_photoProducer)
+            return TakePhotoNativePromise::createAndReject("Internal error"_s);
```
This is an error case here, I would either add ASSERT or release logging.

youennf commented on Source/WebCore/platform/mediastream/mac/AVVideoCaptureSource.mm:
```diff
@@ -448,6 +514,146 @@ static bool isZoomSupported(const Vector<VideoPreset>& presets)
+{
+    Locker lock { m_photoLock };
+
+    if (!m_photoProducer)
```
There should be a photoProducer, should we assert/release_log (above as well).

youennf commented on Source/WebCore/platform/mediastream/mac/AVVideoCaptureSource.mm:
```diff
@@ -448,6 +514,146 @@ static bool isZoomSupported(const Vector<VideoPreset>& presets)
+            ERROR_LOG_IF(loggerPtr(), LOGIDENTIFIER, "unable to add photo output");
+            return nullptr;
+        }
+        [session() addOutput:m_photoOutput.get()];
```
Should we remove the output and destroy m_photoOutput when we are finished with the photo request ?
Or does it have not side effect to keep it even if we only use it once?

youennf commented on Source/WebCore/platform/mediastream/mac/AVVideoCaptureSource.mm:
```diff
@@ -448,6 +514,146 @@ static bool isZoomSupported(const Vector<VideoPreset>& presets)
+
+#if PLATFORM(IOS_FAMILY)
+    auto* photoOutput = this->photoOutput();
+    ASSERT(photoOutput);
```
If canAddOutput return false, this fails. I would return with nullptr and reject the promise in such a case.

eric-carlson committed:
[MediaStream] Add support for ImageCapture.takePhoto
https://bugs.webkit.org/show_bug.cgi?id=262467
rdar://116322637

Reviewed by Youenn Fablet.

Implement ImageCapture.takePhoto in AVVideoCaptureSource and MockRealtimeVideoSource.
We reconfigure the capture source when asked for a photo larger than the current preset,
so it always returns an image at least as large as the requested size, but it doesn't
yet resize the image if the requested size doesn't exactly match the preset.

Taking a photo is an asynchronous operation that can take an arbitrary amount of time, so
we don't do it on the main thread. Because we may need to reconfigure the capture device
to take a photo with the requested settings, we need to prevent the capture device from
being reconfigured by `applyConstraints` while a photo is being generated. To do this we
serialize calls to `takePhoto` and `applyConstraints` so they don't interfere with one
another.

* LayoutTests/fast/mediastream/image-capture-take-photo-expected.txt: Added.
* LayoutTests/fast/mediastream/image-capture-take-photo.html: Added.
* Source/WebCore/Modules/mediastream/ImageCapture.cpp:
(WebCore::ImageCapture::takePhoto):
* Source/WebCore/Modules/mediastream/ImageCapture.h:
* Source/WebCore/Modules/mediastream/ImageCapture.idl:
* Source/WebCore/Modules/mediastream/MediaStreamTrack.cpp:
(WebCore::MediaStreamTrack::stopTrack):
(WebCore::MediaStreamTrack::queueAndProcessSerialAction):
(WebCore::MediaStreamTrack::processNextSerialAction):
(WebCore::MediaStreamTrack::takePhoto):
(WebCore::MediaStreamTrack::applyConstraints):
* Source/WebCore/Modules/mediastream/MediaStreamTrack.h:
* Source/WebCore/PAL/pal/cocoa/AVFoundationSoftLink.h:
* Source/WebCore/PAL/pal/cocoa/AVFoundationSoftLink.mm:
* Source/WebCore/platform/mediastream/MediaStreamTrackPrivate.cpp:
(WebCore::MediaStreamTrackPrivate::takePhoto):
* Source/WebCore/platform/mediastream/MediaStreamTrackPrivate.h:
* Source/WebCore/platform/mediastream/PhotoSettings.h:
(WebCore::operator==):
* Source/WebCore/platform/mediastream/RealtimeMediaSource.cpp:
(WebCore::RealtimeMediaSource::takePhoto):
* Source/WebCore/platform/mediastream/RealtimeMediaSource.h:
* Source/WebCore/platform/mediastream/RealtimeVideoCaptureSource.cpp:
(WebCore::RealtimeVideoCaptureSource::bestSupportedSizeFrameRateAndZoomConsideringObservers):
(WebCore::RealtimeVideoCaptureSource::setSizeFrameRateAndZoom):
(WebCore::RealtimeVideoCaptureSource::takePhotoInternal):
(WebCore::RealtimeVideoCaptureSource::takePhoto):
* Source/WebCore/platform/mediastream/RealtimeVideoCaptureSource.h:
* Source/WebCore/platform/mediastream/mac/AVVideoCaptureSource.h:
* Source/WebCore/platform/mediastream/mac/AVVideoCaptureSource.mm:
(WebCore::toFillLightMode):
(WebCore::toAVCaptureFlashMode):
(WebCore::photoQueue):
(WebCore::AVVideoCaptureSource::~AVVideoCaptureSource):
(WebCore::AVVideoCaptureSource::stopProducingData):
(WebCore::AVVideoCaptureSource::stopSession):
(WebCore::AVVideoCaptureSource::photoOutput):
(WebCore::AVVideoCaptureSource::resolvePendingPhotoRequest):
(WebCore::AVVideoCaptureSource::rejectPendingPhotoRequest):
(WebCore::AVVideoCaptureSource::maxPhotoSizeForCurrentPreset const):
(WebCore::AVVideoCaptureSource::photoConfiguration):
(WebCore::AVVideoCaptureSource::takePhotoInternal):
(WebCore::AVVideoCaptureSource::getPhotoSettings):
(WebCore::AVVideoCaptureSource::captureOutputDidFinishProcessingPhoto):
(-[WebCoreAVVideoCaptureSourceObserver captureOutput:didFinishProcessingPhoto:error:]):
* Source/WebCore/platform/mediastream/mac/MockRealtimeVideoSourceMac.mm:
(WebCore::MockRealtimeVideoSourceMac::updateSampleBuffer):
* Source/WebCore/platform/mock/MockRealtimeVideoSource.cpp:
(WebCore::takePhotoRunLoop):
(WebCore::MockRealtimeVideoSource::DrawingState::fontDescription):
(WebCore::MockRealtimeVideoSource::DrawingState::timeFont):
(WebCore::MockRealtimeVideoSource::DrawingState::bipBopFont):
(WebCore::MockRealtimeVideoSource::DrawingState::statsFont):
(WebCore::MockRealtimeVideoSource::takePhotoInternal):
(WebCore::MockRealtimeVideoSource::invalidateDrawingState):
(WebCore::MockRealtimeVideoSource::drawingState):
(WebCore::MockRealtimeVideoSource::settingsDidChange):
(WebCore::MockRealtimeVideoSource::drawText):
(WebCore::MockRealtimeVideoSource::generatePhoto):
(WebCore::MockRealtimeVideoSource::generateFrameInternal):
(WebCore::MockRealtimeVideoSource::generateFrame):
(WebCore::MockRealtimeVideoSource::imageBuffer):
(WebCore::MockRealtimeVideoSource::imageBufferInternal):
(WebCore::MockRealtimeVideoSource::imageBuffer const): Deleted.
* Source/WebCore/platform/mock/MockRealtimeVideoSource.h:
(WebCore::MockRealtimeVideoSource::DrawingState::DrawingState):
(WebCore::MockRealtimeVideoSource::DrawingState::baseFontSize const):
(WebCore::MockRealtimeVideoSource::DrawingState::statsFontSize const):
* Source/WebKit/GPUProcess/webrtc/RemoteSampleBufferDisplayLayer.h:
* Source/WebKit/UIProcess/Cocoa/UserMediaCaptureManagerProxy.cpp:
(WebKit::UserMediaCaptureManagerProxy::SourceProxy::takePhoto):
(WebKit::UserMediaCaptureManagerProxy::takePhoto):
* Source/WebKit/UIProcess/Cocoa/UserMediaCaptureManagerProxy.h:
* Source/WebKit/UIProcess/Cocoa/UserMediaCaptureManagerProxy.messages.in:
* Source/WebKit/WebProcess/cocoa/RemoteRealtimeMediaSource.cpp:
(WebKit::RemoteRealtimeMediaSource::takePhoto):
* Source/WebKit/WebProcess/cocoa/RemoteRealtimeMediaSource.h:
* Source/WebKit/WebProcess/cocoa/RemoteRealtimeMediaSourceProxy.cpp:
(WebKit::RemoteRealtimeMediaSourceProxy::takePhoto):
* Source/WebKit/WebProcess/cocoa/RemoteRealtimeMediaSourceProxy.h:
ef8342e

Canonical link: https://commits.webkit.org/270621@main
